### PR TITLE
Make following `clerk/doc-url` links usable in interactive mode

### LINKED ISSUE #142: Make links functional when notebooks is displayed via `clerk/show!`
They currently only work in the static app. Can probably make them `clerk-eval` `clerk/show!` given we get the defining notebook file from the link.

### PR DESCRIPTION
**Pull request by zampino**
Allow notebooks to be shown by visiting their path in the browser e.g http://localhost:7777/notebooks/viewers/html.clj

* Make links to a valid notebook path functional in interactive mode
* push URL history in the client when clerk shows a document 

Closes #142.

The above link has examples of links built using `doc-url` for pointing at other notebooks.

Note: this changes the behaviour of `n.c.render/clerk-eval` to not recompute the current notebook by default. Recompute can be opted in via `(clerk-eval form {:recompute? true})`.

### CODE REVIEW

**Comment by zampino** [comment]
Clerk static builds currently do not accept absolute paths (not even pointing inside the content root) because we do expand them via `(fs/glob "." path)` we might address that before changing `n.c.builder/doc-url`.

**Review by mk — COMMENTED** [review]
Besides these comments one thing I ran into testing this is that the exec status bar got stuck. Did you see this as well? Seems to reproduce reliably when navigating back in the browser.

**Comment by mk** [comment]
> An advantage of this approach over https://github.com/nextjournal/clerk/pull/442 is that we can link to sections as in (clerk/doc-url "notebooks/viewer_api.clj#tables").

But that would break in the bundled static case, right? Would be good to support this everywhere.

File: src/nextjournal/clerk/webserver.clj [hunk 1]
```diff
@@ -148,6 +151,19 @@
     (apply swap! nextjournal.clerk.atom/my-state (eval '[update :counter inc]))
     (eval '(nextjournal.clerk/recompute!)))
 
+(declare present+reset! set-status!)
+
+(defn eval-file [file]
+  (main/with-bindings
+    (-> (parser/parse-file {:doc? true} file)
+        (assoc :set-status-fn set-status!)
+        eval/eval-doc)))
```
[the remark this answers is not in the document]
Comment (zampino) [reply src/nextjournal/clerk/webserver.clj:160]: > should we not also do this in other places?

The problem there is that in the usual flow clojure eval is performed within a `clerk/show!` invoked over REPL, and in that context the dynamic binding is already established by clojure.main machinery 

> The REPL itself creates a dynamic binding scope around the eval of each REPL statement with bindings for a hard-coded set of dynamic vars
> -- Alex Miller (https://stackoverflow.com/a/42196114)

The binding scope need be established when eval is performed outside of a REPL invocation like in the thread of Clerk web app handling the response for serving the notebook, I think other spots don't need that.

File: src/nextjournal/clerk/webserver.clj [hunk 2]
```diff
@@ -148,6 +151,19 @@
...
+
+(defn guard [p x] (when (p x) x))
+(defn existing-notebook-path [uri]
+  (when-some [ex (or (guard fs/exists? uri) (guard fs/exists? (subs uri 1)))]
```
[the remark this answers is not in the document]
Comment (zampino) [reply src/nextjournal/clerk/webserver.clj:164]: Yeah, that'd be nicer. but anchor href attribute starting with `//` has its own meaning (reuse current scheme). Shall we explicitely use the full scheme `file://`?

**Comment by zampino** [comment]
> But that would break in the bundled static case

I need to handle fragments explicitly anyway in the overridden doc-url for static builds
https://github.com/nextjournal/clerk/blob/8635abf469b987e1fddf6b4c0869a8b136c06f68/src/nextjournal/clerk/builder.clj#L287-L297
and ignore the fragment in the bundled case. We can address that to fix #449 but I guess we need to use a different reitit router an not the fragment-based one (?). I thought we can try that in a separate pass.

File: src/nextjournal/clerk/webserver.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (zampino) [reply src/nextjournal/clerk/webserver.clj:160]: > don't we want to unify this? 

already unified a bit with `show!`, stopped at a circular dependency between eval and webserver nss when trying to assoc the exec-status fn. Maybe I can overload the first argument to eval-file which is the in-memory cache and assoc the status-fn there or add arity 3.

**Comment by zampino** [comment]
>  the exec status bar got stuck

I didn't see that with Chrome, but tried it now with Safari where exec-status is not working at all for me when loading a notebook path. Maybe Chrome keeps ws connections accept messages while unloading the page.

**Pull request merged**